atlauncher: init at 3.4.28.1

GetPsyched opened the pull request:
###### Description of changes

This is an alternative launcher for Minecraft; similar to a bunch of others like PrismLauncher.
Homepage: https://atlauncher.com/

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

GetPsyched committed:
maintainers: add getpsyched

GetPsyched commented:
I am unsure how do I add a desktop icon for this package. I'm using KDE Plasma and do not see this package on my app menu. I have referred to [this thread](https://discourse.nixos.org/t/create-a-desktop-file-from-script/5126) to no avail.

lucasew commented on pkgs/games/atlauncher/default.nix:
```diff
@@ -0,0 +1,33 @@
+{ fetchurl, jre, lib, makeWrapper, stdenv }:
+
+let
+  version = "3.4.28.1";
```
As you are using a recursive attrset (that rec thing) you dont need to put version in a let expression before.

lucasew commented:
> I am unsure how do I add a desktop icon for this package. I'm using KDE Plasma and do not see this package on my app menu. I have referred to [this thread](https://discourse.nixos.org/t/create-a-desktop-file-from-script/5126) to no avail.

makeDesktopItem + copyDesktopItems

there are many examples over nixpkgs

lucasew commented on pkgs/games/atlauncher/default.nix:
```diff
@@ -0,0 +1,33 @@
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  installPhase = ''
```
`runHook preInstall` at the beginning of installPhase and `runHook postInstall` at the end

it will help you avoid some sneaky bugs

lucasew commented:
BTW in nixpkgs we prefer to package from source code if possible, if not me someone else would say this

GetPsyched commented:
> makeDesktopItem + copyDesktopItems
>
> there are many examples over nixpkgs

I've added a few lines for the same but the icon still does not load. I am assuming I'm doing something blatantly wrong?

> BTW in nixpkgs we prefer to package from source code if possible, if not me someone else would say this

I started off with building from source at first; switched to the jar solely because of errors. I will try again as a refactor after this PR, if that works.

lucasew commented:
The icon doesn't load because it's not found. You can copy a PNG to a specific folder (ex: [1]) or reference an icon downloaded from the Internet by using fetchurl for example.

The first approach is better because in the future an icon pack can override the default icon.

If the project has no icon, you can try using a generic one such as java or another default [2].

[1] https://github.com/NixOS/nixpkgs/blob/6e56adb22679971dfe9343f713786f9cb89b83da/pkgs/applications/office/trilium/desktop.nix#L62
[2] https://specifications.freedesktop.org/icon-naming-spec/icon-naming-spec-latest.html

GetPsyched commented:
Noted for the future. For now, I'm using `fetchurl`.

GetPsyched commented:
Only one thing left in this PR, it's that I need to default the running of this package with `steam-run`. Working on it!

GetPsyched committed:
atlauncher: init at 3.4.28.1

lucasew commented on pkgs/games/atlauncher/default.nix:
```diff
@@ -0,0 +1,59 @@
+      makeWrapper ${jre}/bin/java $out/bin/atlauncher \
+        --add-flags "-jar $src --working-dir=\$HOME/.atlauncher" \
+        --suffix LD_LIBRARY_PATH : "${lib.makeLibraryPath buildInputs}" ${
+            if withSteamRun
```
There is a lib function for that. I think it's `lib.optionalString` or something like that

drupol commented on pkgs/games/atlauncher/default.nix:
```diff
@@ -0,0 +1,57 @@
+{ copyDesktopItems, fetchurl, jre, lib, makeDesktopItem, makeWrapper, stdenv, steam-run, withSteamRun ? true, writeShellScript }:
+
+stdenv.mkDerivation rec {
```
I think we should use the `finalAttrs` pattern here.

Since this is your first PR, let me share a couple of useful links regarding this:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

GetPsyched replied:
Thanks for the links! I've reflected this with the bare-bones understanding I've gotten. Hopefully, it follows the convention.

drupol commented on pkgs/games/atlauncher/default.nix:
```diff
@@ -0,0 +1,57 @@
+
+  src = fetchurl {
+    url = "https://github.com/ATLauncher/ATLauncher/releases/download/v${version}/ATLauncher-${version}.jar";
+    sha256 = "sha256-IIwDMazxUMQ7nGQk/4VEZicgCmCR4oR8UYtO36pCEq4=";
```
```suggestion
    hash = "sha256-IIwDMazxUMQ7nGQk/4VEZicgCmCR4oR8UYtO36pCEq4=";
```

drupol reviewed (approved):
Hi,

This LGTM, however, we usually compile things from the sources instead of downloading the binary release of a software.

Since this is your first contribution, let's accept it as it is, but it would be super nice if you could provide an update compiling it from the sources.

GetPsyched commented:
Hey, drupol. Thanks for considering the binary build; I'll make sure I send another refactor PR building from source in some time.

drupol commented:
Awesome, thanks ! Looking forward to review it!
